
# Introduction to Statistical Analysis of the microbiome data

```{r, eval=TRUE}
library(phyloseq)
library(tidyverse)
library(gt)
```

```{r, echo=FALSE, eval=TRUE, include=TRUE}
klippy::klippy(c('r', 'bash'), position = c('top', 'right'), tooltip_message = 'copy to clipboard', tooltip_success = 'Copied')
```

This section provides a quick introduction to some common analytic methods used to analyze microbiome data. 
In previous sections we worked with microbiome generated data, but ran through some fairly standard protocols that would apply to any Illumina-seq generated data and FASTQ files.

Now we are at the visualisation and statistical analysis phase - and we should start to see the implementation of some very familiar R tools.

Briefly today we will use R to understand more about

- Describing the microbial community composition of a set of samples

- Estimating within- and between-sample diversity

- Identifying differentially abundant taxa


## Important files

We're mostly going to be working with just 3 files now.

- A count table: the number of reads for each unique sequence

- A taxonomy table: the assigned taxonomy for each sequence according to the SILVA database

- A sample file: this is the "metadata" it contains any information *we* have provided about the different samples


If you worked through the entirety of the previous session you will already have the necessary R objects in your workspace. If you didn't we can load them using a `.RData` file. You should see that your Environment is suddenly populated by a number of R objects.

```{r, include=FALSE, eval=TRUE}
load(file="amplicon_dada2_ex.RData")

```

The one thing missing that we also need to load is a sample info object (it tells us meta-data about the origins and conditions of each environmental sample.)


```{r, eval=TRUE}
sample_info_tab <- read.table("sample_info.tsv", header=T, row.names=1,
                   check.names=F, sep="\t")

```


## Phyloseq

Microbiome community analysis works really well with a package called `phyloseq` it allows you to make special R objects that hold the distinct sets of information on ASV abundance, taxonomy and relation to the environmental samples in one R object.

```{r, eval=TRUE}
ps <- phyloseq(otu_table(asv_tab_no_contam, taxa_are_rows=T), 
               sample_data(sample_info_tab), 
               tax_table(asv_tax_no_contam))
ps
```

> **Note**
>
> Since we’ve already used decontam to remove likely contaminants in the previous steps, we’re dropping the “blank” samples from our count table - so now there are 16 samples total. 


<details><summary>**Take a look at each of the three objects: asv_tab_no_contam, sample_data, tax_table. There are just simple lists - can you comfortably work out the information they contain?**</summary>

- `asv_tab_no_contam` A count table: the number of reads for each unique sequence

- `asv_tax_no_contam` A taxonomy table: the assigned taxonomy for each sequence according to the [SILVA](https://www.arb-silva.de/) database

- `sample_info_tab` A sample file: this is the "metadata" it contains any information *we* have provided about the different samples

</details>


## Relative abundance of bacteria by taxonomic Class

We will use some functions provided by `phyloseq` to agglomerate the reads to Class-level and plot the results

```{r, eval=TRUE}
taxonomy <- ps %>% 
  tax_glom(taxrank="class") %>% ### Preserves taxonomy only to Taxonomic level class
  transform_sample_counts(function(x){x/sum(x)})%>% ### this function turns Abundance from raw counts to a relative proportion
  psmelt() %>% ###converts phyloseq object into a tibble 
  filter(Abundance >0.05) ### remove any ASVs at less than 5% abundance

```


Often an early step in many microbiome projects to visualize the relative abundance of organisms at specific taxonomic ranks. 

```{r, eval=TRUE}
taxonomy %>% 
  ggplot(aes(x=Sample, y=Abundance, fill=class))+
  geom_bar(stat="identity")

```
The ability to discriminate between more than say a dozen colors in a single plot is a limitation of the stacked bar plot - so we have filtered out less abundant reads (<5%). We run the risk that we might miss something important here, so its important to visualise the data in different ways. 



There are a total of 10 Bacterial classes (at greater than 5% abundance). 

Let's generate boxplots according to `char` and facet these according to Bacterial `class`. This will allow us to compare how different bacterial classes appear to change in relative abundance between samples. 

Try and produce something similar to the plot below

```{r, echo=FALSE, eval=TRUE}
taxonomy %>% 
  mutate(class=fct_reorder(class, Abundance)) %>% 
  ggplot(aes(x=class, y=Abundance))+
  geom_boxplot()+geom_jitter(aes(color=class), width=0.2)+
  facet_wrap(.~char)+
  theme_classic()+
  theme(axis.text.x=element_text(angle=45, hjust=1), legend.position="none")+
  labs(x="Bacterial class", y="Relative Abundance")
```

Think about what you can infer from these plots?

## Alpha diversity - community richness

This is a common ecological term, and one that applies just as well when we consider microbial ecosystems.

What is the mean species diversity in the different sample sites?

```{r, eval=FALSE}
ps_richness <- estimate_richness(ps)
ps_richness
```

```{r, echo=FALSE, eval=TRUE}
ps_richness <- estimate_richness(ps)

```

This command generates a number of different diversity indices. 
Firstly we can take this information and use `join` functions to attach it back onto our sample *metadata*

```{r, eval=FALSE}
sample_info_tab$Sample <- rownames(sample_info_tab)

ps_richness$Sample <- rownames(ps_richness)

full_richness <- left_join(sample_info_tab, ps_richness, "Sample")
full_richness 
```

```{r, echo=FALSE, eval=TRUE}
sample_info_tab$Sample <- rownames(sample_info_tab)

ps_richness$Sample <- rownames(ps_richness)

full_richness <- left_join(sample_info_tab, ps_richness, "Sample")
full_richness %>% 
  gt()
```


### Analyse and plot

We now have a tidy data frame with sample information and a variety of diversity indices. 

- Using the Shannon index as the measure of diversity - can you construct a figure to show the diversity differences between treatments?

- Again using the Shannon index - can you construct a `general linear model` to work out which treatments/effects might be having a significant effect on diversity?



## Beta diversity

### Heatmaps

Heatmaps are basically false colour images where cells in the matrix with high relative values are coloured differently from those with low relative values. Heatmaps can range from very simple blocks of colour with lists along 2 sides, or they can include information about hierarchical clustering, and/or values of other covariates of interest.

We will make a simple heatmap here, but a common addition might be a dendrogram - a phylogenetic tree to indicate how closely related different 

```{r, eval=TRUE}
ps_top20 <- prune_taxa(names(sort(taxa_sums(ps),TRUE)[1:20]), ps) ## subset to include only the top 20 most abundant ASVs

plot_heatmap(ps_top20,"MDS",distance = "bray", ## here we are setting up the parameters for the heat map
             sample.label="char", ## set the x axis to display by sample
             low="purple",   ## set the color for low abundance OTUs
             high="darkorange") ## set the color for high abundance OTUs

```


### Multidimensional analysis

Often in ecological research, we are interested not only in comparing univariate descriptors of communities, like diversity (above), but also in how the constituent species — or the composition — changes from one community to the next.

One common tool to do this is non-metric multidimensional scaling, or NMDS. The goal of NMDS is to collapse information from multiple dimensions (e.g, from multiple communities, sites, etc.) into just a few, so that they can be visualized and interpreted.

Consider a single axis representing the abundance of a single species. Along this axis, we can plot the communities in which this species appears, based on its abundance within each.

Now consider a second axis of abundance, representing another species. We can now plot each community along the two axes (Species 1 and Species 2).

Keep going and imagine how many axes we would have to produce!!!!

The goal of NMDS is to represent the original position of communities in multidimensional space as accurately as possible using a reduced number of dimensions that can be easily plotted and visualized (and to spare your thinker).

NMDS does not use the absolute abundances of species in communities, but rather their rank orders. The use of ranks omits some of the issues associated with using absolute distance (e.g., sensitivity to transformation), and as a result is a flexible technique


```{r, eval=TRUE}
    ### Ordination using built in functions in phyoseq (calls vegan)    
ord.nmds.bray <- ordinate(ps, method="NMDS",k=2, distance="bray")
```

You should see each iteration of the NMDS until a solution is reached (i.e., stress was minimized after some number of reconfigurations of the points in 2 dimensions). You can increase the number of default iterations using the argument trymax=. which may help alleviate issues of non-convergence. If high stress is your problem, increasing the number of dimensions to k=3 might also help.


Often in ecological research, we are interested not only in comparing univariate descriptors of communities, like diversity (above), but also in how the constituent species — or the composition — changes from one community to the next.

One common tool to do this is non-metric multidimensional scaling, or NMDS. The goal of NMDS is to collapse information from multiple dimensions (e.g, from multiple communities, sites, etc.) into just a few, so that they can be visualized and interpreted.

Consider a single axis representing the abundance of a single species. Along this axis, we can plot the communities in which this species appears, based on its abundance within each.

Now consider a second axis of abundance, representing another species. We can now plot each community along the two axes (Species 1 and Species 2).

Keep going and imagine how many axes we would have to produce!!!!

The goal of NMDS is to represent the original position of communities in multidimensional space as accurately as possible using a reduced number of dimensions that can be easily plotted and visualized (and to spare your thinker).

NMDS does not use the absolute abundances of species in communities, but rather their rank orders. The use of ranks omits some of the issues associated with using absolute distance (e.g., sensitivity to transformation), and as a result is a flexible technique



We can use the function plot.ordination to use our NMDS scaling to position our samples on a ggplot

```{r, eval=TRUE}
ord1<-plot_ordination(ps, ord.nmds.bray, color="char", title="Bray NMDS")
ord1
```


And we can add customisations to improve the plot

```{r, eval=TRUE, warning=FALSE, message=FALSE}
ord1 + stat_ellipse(type="norm",
                    geom="polygon",
                    aes(fill=char),
                    alpha=0.3,
                    linetype="dashed") +
     geom_point(pch=21, aes(fill=char), colour="white", stroke=0.5, size=3)+
        theme_classic()
```


## Question

We have now produced several different visualisations and analyses of the microbiome communities from these deep sea samples. 
What do you think are the most important or interesting findings here?


## Summary

This is a *very* brief introduction to using **big** data. Hopefully you can see that many of the command line tools, data cleaning and analysis and visualisation you have been practising prepare you for Big Data Biology.

In this example, in order to have generated and processed this data, you would have needed to know 

- Sampling techniques
- Molecular lab skills
- Data analysis and coding knowledge
- Ecological skills (community analysis)

Biology is increasingly interdisciplinary - and the skills you learn in one module should be transferable to other things you do. 

And remember at the heart of everything is Data! It doesn't matter what you are studying or why - good data skills are fundamental.

Thank you! 
