<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Introduction to Statistical Analysis of the microbiome data | Introduction to Bioinformatics - 5023Y</title>
  <meta name="description" content="This is very brief introduction to some of the basic skills required to handling sequencing data for Bioinformatics - with one reproducible example analysing microbiome sequencing data" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Introduction to Statistical Analysis of the microbiome data | Introduction to Bioinformatics - 5023Y" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is very brief introduction to some of the basic skills required to handling sequencing data for Bioinformatics - with one reproducible example analysing microbiome sequencing data" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Introduction to Statistical Analysis of the microbiome data | Introduction to Bioinformatics - 5023Y" />
  
  <meta name="twitter:description" content="This is very brief introduction to some of the basic skills required to handling sequencing data for Bioinformatics - with one reproducible example analysing microbiome sequencing data" />
  

<meta name="author" content="Philip Leftwich" />


<meta name="date" content="2021-08-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="ngs-sequence-analysis.html"/>
<link rel="next" href="applications.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="libs/bsTable-3.3.7/bootstrapTable.min.css" rel="stylesheet" />
<script src="libs/bsTable-3.3.7/bootstrapTable.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="Unix.html"><a href="Unix.html"><i class="fa fa-check"></i><b>2</b> Unix</a>
<ul>
<li class="chapter" data-level="2.1" data-path="Unix.html"><a href="Unix.html#what-is-unixlinux"><i class="fa fa-check"></i><b>2.1</b> What is Unix/Linux?</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="Unix.html"><a href="Unix.html#some-terms"><i class="fa fa-check"></i><b>2.1.1</b> Some terms</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="Unix.html"><a href="Unix.html#why-learn-unix"><i class="fa fa-check"></i><b>2.2</b> Why Learn Unix?</a></li>
<li class="chapter" data-level="2.3" data-path="Unix.html"><a href="Unix.html#getting-started"><i class="fa fa-check"></i><b>2.3</b> Getting started</a></li>
<li class="chapter" data-level="2.4" data-path="Unix.html"><a href="Unix.html#a-few-foundational-rules"><i class="fa fa-check"></i><b>2.4</b> A few foundational rules</a></li>
<li class="chapter" data-level="2.5" data-path="Unix.html"><a href="Unix.html#lets-get-started"><i class="fa fa-check"></i><b>2.5</b> Let’s get started</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="Unix.html"><a href="Unix.html#downloading-data"><i class="fa fa-check"></i><b>2.5.1</b> Downloading data</a></li>
<li class="chapter" data-level="2.5.2" data-path="Unix.html"><a href="Unix.html#more-functions"><i class="fa fa-check"></i><b>2.5.2</b> More functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="Unix.html"><a href="Unix.html#unix-file-structure"><i class="fa fa-check"></i><b>2.6</b> Unix File Structure</a></li>
<li class="chapter" data-level="2.7" data-path="Unix.html"><a href="Unix.html#absolute-vs-relative-file-paths"><i class="fa fa-check"></i><b>2.7</b> Absolute vs relative file paths</a></li>
<li class="chapter" data-level="2.8" data-path="Unix.html"><a href="Unix.html#moving-around"><i class="fa fa-check"></i><b>2.8</b> Moving around</a></li>
<li class="chapter" data-level="2.9" data-path="Unix.html"><a href="Unix.html#summary"><i class="fa fa-check"></i><b>2.9</b> Summary</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="Unix.html"><a href="Unix.html#terms"><i class="fa fa-check"></i><b>2.9.1</b> Terms</a></li>
<li class="chapter" data-level="2.9.2" data-path="Unix.html"><a href="Unix.html#commands"><i class="fa fa-check"></i><b>2.9.2</b> Commands</a></li>
<li class="chapter" data-level="2.9.3" data-path="Unix.html"><a href="Unix.html#special-characters"><i class="fa fa-check"></i><b>2.9.3</b> Special characters</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="Unix.html"><a href="Unix.html#summary-1"><i class="fa fa-check"></i><b>2.10</b> Summary</a></li>
<li class="chapter" data-level="2.11" data-path="Unix.html"><a href="Unix.html#stretch-yourself---optional-extras-to-try-a-couple-of-other-skills"><i class="fa fa-check"></i><b>2.11</b> Stretch yourself - optional extras to try a couple of other skills</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="Unix.html"><a href="Unix.html#creation"><i class="fa fa-check"></i><b>2.11.1</b> Creation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html"><i class="fa fa-check"></i><b>3</b> NGS sequence analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#background-on-high-throughput-sequencing"><i class="fa fa-check"></i><b>3.1</b> Background on high throughput sequencing</a></li>
<li class="chapter" data-level="3.2" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#some-terms-1"><i class="fa fa-check"></i><b>3.2</b> Some terms</a></li>
<li class="chapter" data-level="3.3" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#the-data"><i class="fa fa-check"></i><b>3.3</b> The data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#fastq"><i class="fa fa-check"></i><b>3.3.1</b> FASTQ?</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#the-pipeline"><i class="fa fa-check"></i><b>3.4</b> The Pipeline</a></li>
<li class="chapter" data-level="3.5" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#removing-primers"><i class="fa fa-check"></i><b>3.5</b> Removing Primers</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#loops"><i class="fa fa-check"></i><b>3.5.1</b> Loops</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#dada2"><i class="fa fa-check"></i><b>3.6</b> DADA2</a></li>
<li class="chapter" data-level="3.7" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#quality-trimmingfiltering"><i class="fa fa-check"></i><b>3.7</b> Quality trimming/filtering</a></li>
<li class="chapter" data-level="3.8" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#dereplication"><i class="fa fa-check"></i><b>3.8</b> Dereplication</a></li>
<li class="chapter" data-level="3.9" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#asvs"><i class="fa fa-check"></i><b>3.9</b> ASV’s</a></li>
<li class="chapter" data-level="3.10" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#merging-reads"><i class="fa fa-check"></i><b>3.10</b> Merging reads</a></li>
<li class="chapter" data-level="3.11" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#count-table"><i class="fa fa-check"></i><b>3.11</b> Count table</a></li>
<li class="chapter" data-level="3.12" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#overview"><i class="fa fa-check"></i><b>3.12</b> Overview</a></li>
<li class="chapter" data-level="3.13" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#assign-taxonomy"><i class="fa fa-check"></i><b>3.13</b> Assign taxonomy</a></li>
<li class="chapter" data-level="3.14" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#standard-goods"><i class="fa fa-check"></i><b>3.14</b> Standard goods</a></li>
<li class="chapter" data-level="3.15" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#functions-list"><i class="fa fa-check"></i><b>3.15</b> Functions list</a></li>
<li class="chapter" data-level="3.16" data-path="ngs-sequence-analysis.html"><a href="ngs-sequence-analysis.html#summary-2"><i class="fa fa-check"></i><b>3.16</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html"><i class="fa fa-check"></i><b>4</b> Introduction to Statistical Analysis of the microbiome data</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#important-files"><i class="fa fa-check"></i><b>4.1</b> Important files</a></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#phyloseq"><i class="fa fa-check"></i><b>4.2</b> Phyloseq</a></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#relative-abundance-of-bacteria-by-taxonomic-class"><i class="fa fa-check"></i><b>4.3</b> Relative abundance of bacteria by taxonomic Class</a></li>
<li class="chapter" data-level="4.4" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#alpha-diversity---community-richness"><i class="fa fa-check"></i><b>4.4</b> Alpha diversity - community richness</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#analyse-and-plot"><i class="fa fa-check"></i><b>4.4.1</b> Analyse and plot</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#beta-diversity"><i class="fa fa-check"></i><b>4.5</b> Beta diversity</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#heatmaps"><i class="fa fa-check"></i><b>4.5.1</b> Heatmaps</a></li>
<li class="chapter" data-level="4.5.2" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#multidimensional-analysis"><i class="fa fa-check"></i><b>4.5.2</b> Multidimensional analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#question"><i class="fa fa-check"></i><b>4.6</b> Question</a></li>
<li class="chapter" data-level="4.7" data-path="introduction-to-statistical-analysis-of-the-microbiome-data.html"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#summary-3"><i class="fa fa-check"></i><b>4.7</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="applications.html"><a href="applications.html"><i class="fa fa-check"></i><b>5</b> Applications</a>
<ul>
<li class="chapter" data-level="5.1" data-path="applications.html"><a href="applications.html#example-one"><i class="fa fa-check"></i><b>5.1</b> Example one</a></li>
<li class="chapter" data-level="5.2" data-path="applications.html"><a href="applications.html#example-two"><i class="fa fa-check"></i><b>5.2</b> Example two</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="final-words.html"><a href="final-words.html"><i class="fa fa-check"></i><b>6</b> Final Words</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Bioinformatics - 5023Y</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-statistical-analysis-of-the-microbiome-data" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Introduction to Statistical Analysis of the microbiome data</h1>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb48-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb48-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gt)</span></code></pre></div>
<script>
  addClassKlippyTo("pre.r, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'copy to clipboard', 'Copied');
</script>
<p>This section provides a quick introduction to some common analytic methods used to analyze microbiome data.
In previous sections we worked with microbiome generated data, but ran through some fairly standard protocols that would apply to any Illumina-seq generated data and FASTQ files.</p>
<p>Now we are at the visualisation and statistical analysis phase - and we should start to see the implementation of some very familiar R tools.</p>
<p>Briefly today we will use R to understand more about</p>
<ul>
<li><p>Describing the microbial community composition of a set of samples</p></li>
<li><p>Estimating within- and between-sample diversity</p></li>
<li><p>Identifying differentially abundant taxa</p></li>
</ul>
<div id="important-files" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Important files</h2>
<p>We’re mostly going to be working with just 3 files now.</p>
<ul>
<li><p>A count table: the number of reads for each unique sequence</p></li>
<li><p>A taxonomy table: the assigned taxonomy for each sequence according to the SILVA database</p></li>
<li><p>A sample file: this is the “metadata” it contains any information <em>we</em> have provided about the different samples</p></li>
</ul>
<p>If you worked through the entirety of the previous session you will already have the necessary R objects in your workspace. If you didn’t we can load them using a <code>.RData</code> file. You should see that your Environment is suddenly populated by a number of R objects.</p>
<p>The one thing missing that we also need to load is a sample info object (it tells us meta-data about the origins and conditions of each environmental sample.)</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb49-1" aria-hidden="true" tabindex="-1"></a>sample_info_tab <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;sample_info.tsv&quot;</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="dv">1</span>,</span>
<span id="cb49-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb49-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">check.names=</span>F, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="phyloseq" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Phyloseq</h2>
<p>Microbiome community analysis works really well with a package called <code>phyloseq</code> it allows you to make special R objects that hold the distinct sets of information on ASV abundance, taxonomy and relation to the environmental samples in one R object.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb50-1" aria-hidden="true" tabindex="-1"></a>ps <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(<span class="fu">otu_table</span>(asv_tab_no_contam, <span class="at">taxa_are_rows=</span>T), </span>
<span id="cb50-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb50-2" aria-hidden="true" tabindex="-1"></a>               <span class="fu">sample_data</span>(sample_info_tab), </span>
<span id="cb50-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb50-3" aria-hidden="true" tabindex="-1"></a>               <span class="fu">tax_table</span>(asv_tax_no_contam))</span>
<span id="cb50-4"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb50-4" aria-hidden="true" tabindex="-1"></a>ps</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 2498 taxa and 16 samples ]
## sample_data() Sample Data:       [ 16 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 2498 taxa by 7 taxonomic ranks ]</code></pre>
<blockquote>
<p><strong>Note</strong></p>
<p>Since we’ve already used decontam to remove likely contaminants in the previous steps, we’re dropping the “blank” samples from our count table - so now there are 16 samples total.</p>
</blockquote>
<details>
<summary>
<strong>Take a look at each of the three objects: asv_tab_no_contam, sample_data, tax_table. There are just simple lists - can you comfortably work out the information they contain?</strong>
</summary>
<ul>
<li><p><code>asv_tab_no_contam</code> A count table: the number of reads for each unique sequence</p></li>
<li><p><code>asv_tax_no_contam</code> A taxonomy table: the assigned taxonomy for each sequence according to the <a href="https://www.arb-silva.de/">SILVA</a> database</p></li>
<li><p><code>sample_info_tab</code> A sample file: this is the “metadata” it contains any information <em>we</em> have provided about the different samples</p></li>
</ul>
</details>
</div>
<div id="relative-abundance-of-bacteria-by-taxonomic-class" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Relative abundance of bacteria by taxonomic Class</h2>
<p>We will use some functions provided by <code>phyloseq</code> to agglomerate the reads to Class-level and plot the results</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb52-1" aria-hidden="true" tabindex="-1"></a>taxonomy <span class="ot">&lt;-</span> ps <span class="sc">%&gt;%</span> </span>
<span id="cb52-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_glom</span>(<span class="at">taxrank=</span><span class="st">&quot;class&quot;</span>) <span class="sc">%&gt;%</span> <span class="do">### Preserves taxonomy only to Taxonomic level class</span></span>
<span id="cb52-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(<span class="cf">function</span>(x){x<span class="sc">/</span><span class="fu">sum</span>(x)})<span class="sc">%&gt;%</span> <span class="do">### this function turns Abundance from raw counts to a relative proportion</span></span>
<span id="cb52-4"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">psmelt</span>() <span class="sc">%&gt;%</span> <span class="do">###converts phyloseq object into a tibble </span></span>
<span id="cb52-5"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Abundance <span class="sc">&gt;</span><span class="fl">0.05</span>) <span class="do">### remove any ASVs at less than 5% abundance</span></span></code></pre></div>
<p>Often an early step in many microbiome projects to visualize the relative abundance of organisms at specific taxonomic ranks.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb53-1" aria-hidden="true" tabindex="-1"></a>taxonomy <span class="sc">%&gt;%</span> </span>
<span id="cb53-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>class))<span class="sc">+</span></span>
<span id="cb53-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-67-1.png" width="672" />
The ability to discriminate between more than say a dozen colors in a single plot is a limitation of the stacked bar plot - so we have filtered out less abundant reads (&lt;5%). We run the risk that we might miss something important here, so its important to visualise the data in different ways.</p>
<p>There are a total of 10 Bacterial classes (at greater than 5% abundance).</p>
<p>Let’s generate boxplots according to <code>char</code> and facet these according to Bacterial <code>class</code>. This will allow us to compare how different bacterial classes appear to change in relative abundance between samples.</p>
<p>Try and produce something similar to the plot below</p>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
<p>Think about what you can infer from these plots?</p>
</div>
<div id="alpha-diversity---community-richness" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> Alpha diversity - community richness</h2>
<p>This is a common ecological term, and one that applies just as well when we consider microbial ecosystems.</p>
<p>What is the mean species diversity in the different sample sites?</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb54-1" aria-hidden="true" tabindex="-1"></a>ps_richness <span class="ot">&lt;-</span> <span class="fu">estimate_richness</span>(ps)</span>
<span id="cb54-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb54-2" aria-hidden="true" tabindex="-1"></a>ps_richness</span></code></pre></div>
<p>This command generates a number of different diversity indices.
Firstly we can take this information and use <code>join</code> functions to attach it back onto our sample <em>metadata</em></p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-1" aria-hidden="true" tabindex="-1"></a>sample_info_tab<span class="sc">$</span>Sample <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sample_info_tab)</span>
<span id="cb55-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-3" aria-hidden="true" tabindex="-1"></a>ps_richness<span class="sc">$</span>Sample <span class="ot">&lt;-</span> <span class="fu">rownames</span>(ps_richness)</span>
<span id="cb55-4"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-5" aria-hidden="true" tabindex="-1"></a>full_richness <span class="ot">&lt;-</span> <span class="fu">left_join</span>(sample_info_tab, ps_richness, <span class="st">&quot;Sample&quot;</span>)</span>
<span id="cb55-6"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb55-6" aria-hidden="true" tabindex="-1"></a>full_richness </span></code></pre></div>
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#bqomojtylo .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#bqomojtylo .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bqomojtylo .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#bqomojtylo .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 4px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#bqomojtylo .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bqomojtylo .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bqomojtylo .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#bqomojtylo .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#bqomojtylo .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#bqomojtylo .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#bqomojtylo .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#bqomojtylo .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#bqomojtylo .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#bqomojtylo .gt_from_md > :first-child {
  margin-top: 0;
}

#bqomojtylo .gt_from_md > :last-child {
  margin-bottom: 0;
}

#bqomojtylo .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#bqomojtylo .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#bqomojtylo .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bqomojtylo .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#bqomojtylo .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bqomojtylo .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#bqomojtylo .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#bqomojtylo .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bqomojtylo .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bqomojtylo .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#bqomojtylo .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bqomojtylo .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#bqomojtylo .gt_left {
  text-align: left;
}

#bqomojtylo .gt_center {
  text-align: center;
}

#bqomojtylo .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#bqomojtylo .gt_font_normal {
  font-weight: normal;
}

#bqomojtylo .gt_font_bold {
  font-weight: bold;
}

#bqomojtylo .gt_font_italic {
  font-style: italic;
}

#bqomojtylo .gt_super {
  font-size: 65%;
}

#bqomojtylo .gt_footnote_marks {
  font-style: italic;
  font-size: 65%;
}
</style>
<div id="bqomojtylo" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;"><table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">temp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">type</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">char</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">color</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Sample</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Observed</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Chao1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">se.chao1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">ACE</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">se.ACE</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Shannon</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Simpson</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">InvSimpson</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Fisher</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">2.0</td>
      <td class="gt_row gt_center">water</td>
      <td class="gt_row gt_center">water</td>
      <td class="gt_row gt_center">blue</td>
      <td class="gt_row gt_left">BW1</td>
      <td class="gt_row gt_right">61</td>
      <td class="gt_row gt_right">62.5000</td>
      <td class="gt_row gt_right">2.5911387</td>
      <td class="gt_row gt_right">61.88289</td>
      <td class="gt_row gt_right">3.569938</td>
      <td class="gt_row gt_right">3.576981</td>
      <td class="gt_row gt_right">0.9585698</td>
      <td class="gt_row gt_right">24.13700</td>
      <td class="gt_row gt_right">12.04096</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.0</td>
      <td class="gt_row gt_center">water</td>
      <td class="gt_row gt_center">water</td>
      <td class="gt_row gt_center">blue</td>
      <td class="gt_row gt_left">BW2</td>
      <td class="gt_row gt_right">204</td>
      <td class="gt_row gt_right">205.9091</td>
      <td class="gt_row gt_right">2.1571603</td>
      <td class="gt_row gt_right">205.71458</td>
      <td class="gt_row gt_right">7.069400</td>
      <td class="gt_row gt_right">4.436499</td>
      <td class="gt_row gt_right">0.9750331</td>
      <td class="gt_row gt_right">40.05305</td>
      <td class="gt_row gt_right">43.45043</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">13.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">glassy</td>
      <td class="gt_row gt_center">black</td>
      <td class="gt_row gt_left">R10</td>
      <td class="gt_row gt_right">383</td>
      <td class="gt_row gt_right">385.4375</td>
      <td class="gt_row gt_right">2.1449492</td>
      <td class="gt_row gt_right">386.66260</td>
      <td class="gt_row gt_right">9.717693</td>
      <td class="gt_row gt_right">5.079826</td>
      <td class="gt_row gt_right">0.9872712</td>
      <td class="gt_row gt_right">78.56183</td>
      <td class="gt_row gt_right">81.57801</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">7.3</td>
      <td class="gt_row gt_center">biofilm</td>
      <td class="gt_row gt_center">biofilm</td>
      <td class="gt_row gt_center">darkgreen</td>
      <td class="gt_row gt_left">R11BF</td>
      <td class="gt_row gt_right">144</td>
      <td class="gt_row gt_right">145.1538</td>
      <td class="gt_row gt_right">1.5236935</td>
      <td class="gt_row gt_right">146.00669</td>
      <td class="gt_row gt_right">5.980036</td>
      <td class="gt_row gt_right">3.309788</td>
      <td class="gt_row gt_right">0.9044015</td>
      <td class="gt_row gt_right">10.46042</td>
      <td class="gt_row gt_right">25.68876</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">7.3</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">glassy</td>
      <td class="gt_row gt_center">black</td>
      <td class="gt_row gt_left">R11</td>
      <td class="gt_row gt_right">281</td>
      <td class="gt_row gt_right">284.4737</td>
      <td class="gt_row gt_right">2.9104269</td>
      <td class="gt_row gt_right">284.57048</td>
      <td class="gt_row gt_right">8.337302</td>
      <td class="gt_row gt_right">4.812376</td>
      <td class="gt_row gt_right">0.9851114</td>
      <td class="gt_row gt_right">67.16529</td>
      <td class="gt_row gt_right">59.63308</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">NA</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R12</td>
      <td class="gt_row gt_right">524</td>
      <td class="gt_row gt_right">525.8947</td>
      <td class="gt_row gt_right">1.9684441</td>
      <td class="gt_row gt_right">525.94449</td>
      <td class="gt_row gt_right">11.362166</td>
      <td class="gt_row gt_right">5.789071</td>
      <td class="gt_row gt_right">0.9956479</td>
      <td class="gt_row gt_right">229.77212</td>
      <td class="gt_row gt_right">115.57917</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">8.6</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R1A</td>
      <td class="gt_row gt_right">497</td>
      <td class="gt_row gt_right">501.2500</td>
      <td class="gt_row gt_right">2.9984998</td>
      <td class="gt_row gt_right">502.32876</td>
      <td class="gt_row gt_right">11.107301</td>
      <td class="gt_row gt_right">5.666947</td>
      <td class="gt_row gt_right">0.9945360</td>
      <td class="gt_row gt_right">183.01554</td>
      <td class="gt_row gt_right">115.04680</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">8.6</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R1B</td>
      <td class="gt_row gt_right">577</td>
      <td class="gt_row gt_right">579.6000</td>
      <td class="gt_row gt_right">2.2526890</td>
      <td class="gt_row gt_right">580.24471</td>
      <td class="gt_row gt_right">11.965009</td>
      <td class="gt_row gt_right">5.812676</td>
      <td class="gt_row gt_right">0.9953007</td>
      <td class="gt_row gt_right">212.79743</td>
      <td class="gt_row gt_right">129.04718</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">8.6</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R2</td>
      <td class="gt_row gt_right">608</td>
      <td class="gt_row gt_right">609.4062</td>
      <td class="gt_row gt_right">1.5321223</td>
      <td class="gt_row gt_right">610.17688</td>
      <td class="gt_row gt_right">12.296746</td>
      <td class="gt_row gt_right">5.850938</td>
      <td class="gt_row gt_right">0.9955702</td>
      <td class="gt_row gt_right">225.74143</td>
      <td class="gt_row gt_right">133.85232</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">12.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R3</td>
      <td class="gt_row gt_right">566</td>
      <td class="gt_row gt_right">567.5517</td>
      <td class="gt_row gt_right">1.6446664</td>
      <td class="gt_row gt_right">568.36545</td>
      <td class="gt_row gt_right">11.819663</td>
      <td class="gt_row gt_right">5.788769</td>
      <td class="gt_row gt_right">0.9953827</td>
      <td class="gt_row gt_right">216.57487</td>
      <td class="gt_row gt_right">122.16253</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">12.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R4</td>
      <td class="gt_row gt_right">643</td>
      <td class="gt_row gt_right">644.2857</td>
      <td class="gt_row gt_right">1.4737632</td>
      <td class="gt_row gt_right">644.75628</td>
      <td class="gt_row gt_right">12.645710</td>
      <td class="gt_row gt_right">5.718854</td>
      <td class="gt_row gt_right">0.9940234</td>
      <td class="gt_row gt_right">167.31991</td>
      <td class="gt_row gt_right">137.19651</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">12.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R5</td>
      <td class="gt_row gt_right">624</td>
      <td class="gt_row gt_right">624.2069</td>
      <td class="gt_row gt_right">0.5163104</td>
      <td class="gt_row gt_right">624.78554</td>
      <td class="gt_row gt_right">12.367463</td>
      <td class="gt_row gt_right">5.966462</td>
      <td class="gt_row gt_right">0.9962834</td>
      <td class="gt_row gt_right">269.06381</td>
      <td class="gt_row gt_right">137.21126</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">12.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">altered</td>
      <td class="gt_row gt_center">chocolate4</td>
      <td class="gt_row gt_left">R6</td>
      <td class="gt_row gt_right">523</td>
      <td class="gt_row gt_right">527.2857</td>
      <td class="gt_row gt_right">3.1338410</td>
      <td class="gt_row gt_right">527.52112</td>
      <td class="gt_row gt_right">11.340777</td>
      <td class="gt_row gt_right">5.711816</td>
      <td class="gt_row gt_right">0.9948403</td>
      <td class="gt_row gt_right">193.80933</td>
      <td class="gt_row gt_right">115.90757</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">NA</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">carbonate</td>
      <td class="gt_row gt_center">darkkhaki</td>
      <td class="gt_row gt_left">R7</td>
      <td class="gt_row gt_right">435</td>
      <td class="gt_row gt_right">436.7838</td>
      <td class="gt_row gt_right">1.7397113</td>
      <td class="gt_row gt_right">438.10995</td>
      <td class="gt_row gt_right">10.011218</td>
      <td class="gt_row gt_right">5.557851</td>
      <td class="gt_row gt_right">0.9939838</td>
      <td class="gt_row gt_right">166.21652</td>
      <td class="gt_row gt_right">110.12229</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">13.5</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">glassy</td>
      <td class="gt_row gt_center">black</td>
      <td class="gt_row gt_left">R8</td>
      <td class="gt_row gt_right">424</td>
      <td class="gt_row gt_right">428.1379</td>
      <td class="gt_row gt_right">3.0460370</td>
      <td class="gt_row gt_right">428.53605</td>
      <td class="gt_row gt_right">10.245312</td>
      <td class="gt_row gt_right">5.295582</td>
      <td class="gt_row gt_right">0.9910943</td>
      <td class="gt_row gt_right">112.28782</td>
      <td class="gt_row gt_right">91.16232</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">13.7</td>
      <td class="gt_row gt_center">rock</td>
      <td class="gt_row gt_center">glassy</td>
      <td class="gt_row gt_center">black</td>
      <td class="gt_row gt_left">R9</td>
      <td class="gt_row gt_right">296</td>
      <td class="gt_row gt_right">300.5652</td>
      <td class="gt_row gt_right">3.3753169</td>
      <td class="gt_row gt_right">300.44067</td>
      <td class="gt_row gt_right">8.397187</td>
      <td class="gt_row gt_right">4.348445</td>
      <td class="gt_row gt_right">0.9517400</td>
      <td class="gt_row gt_right">20.72112</td>
      <td class="gt_row gt_right">63.39469</td>
    </tr>
  </tbody>
  
  
</table></div>
<div id="analyse-and-plot" class="section level3" number="4.4.1">
<h3><span class="header-section-number">4.4.1</span> Analyse and plot</h3>
<p>We now have a tidy data frame with sample information and a variety of diversity indices.</p>
<ul>
<li><p>Using the Shannon index as the measure of diversity - can you construct a figure to show the diversity differences between treatments?</p></li>
<li><p>Again using the Shannon index - can you construct a <code>general linear model</code> to work out which treatments/effects might be having a significant effect on diversity?</p></li>
</ul>
</div>
</div>
<div id="beta-diversity" class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> Beta diversity</h2>
<div id="heatmaps" class="section level3" number="4.5.1">
<h3><span class="header-section-number">4.5.1</span> Heatmaps</h3>
<p>Heatmaps are basically false colour images where cells in the matrix with high relative values are coloured differently from those with low relative values. Heatmaps can range from very simple blocks of colour with lists along 2 sides, or they can include information about hierarchical clustering, and/or values of other covariates of interest.</p>
<p>We will make a simple heatmap here, but a common addition might be a dendrogram - a phylogenetic tree to indicate how closely related different</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-1" aria-hidden="true" tabindex="-1"></a>ps_top20 <span class="ot">&lt;-</span> <span class="fu">prune_taxa</span>(<span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">taxa_sums</span>(ps),<span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]), ps) <span class="do">## subset to include only the top 20 most abundant ASVs</span></span>
<span id="cb56-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_heatmap</span>(ps_top20,<span class="st">&quot;MDS&quot;</span>,<span class="at">distance =</span> <span class="st">&quot;bray&quot;</span>, <span class="do">## here we are setting up the parameters for the heat map</span></span>
<span id="cb56-4"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">sample.label=</span><span class="st">&quot;char&quot;</span>, <span class="do">## set the x axis to display by sample</span></span>
<span id="cb56-5"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">low=</span><span class="st">&quot;purple&quot;</span>,   <span class="do">## set the color for low abundance OTUs</span></span>
<span id="cb56-6"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb56-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">high=</span><span class="st">&quot;darkorange&quot;</span>) <span class="do">## set the color for high abundance OTUs</span></span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-73-1.png" width="672" /></p>
</div>
<div id="multidimensional-analysis" class="section level3" number="4.5.2">
<h3><span class="header-section-number">4.5.2</span> Multidimensional analysis</h3>
<p>Often in ecological research, we are interested not only in comparing univariate descriptors of communities, like diversity (above), but also in how the constituent species — or the composition — changes from one community to the next.</p>
<p>One common tool to do this is non-metric multidimensional scaling, or NMDS. The goal of NMDS is to collapse information from multiple dimensions (e.g, from multiple communities, sites, etc.) into just a few, so that they can be visualized and interpreted.</p>
<p>Consider a single axis representing the abundance of a single species. Along this axis, we can plot the communities in which this species appears, based on its abundance within each.</p>
<p>Now consider a second axis of abundance, representing another species. We can now plot each community along the two axes (Species 1 and Species 2).</p>
<p>Keep going and imagine how many axes we would have to produce!!!!</p>
<p>The goal of NMDS is to represent the original position of communities in multidimensional space as accurately as possible using a reduced number of dimensions that can be easily plotted and visualized (and to spare your thinker).</p>
<p>NMDS does not use the absolute abundances of species in communities, but rather their rank orders. The use of ranks omits some of the issues associated with using absolute distance (e.g., sensitivity to transformation), and as a result is a flexible technique</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb57-1" aria-hidden="true" tabindex="-1"></a>    <span class="do">### Ordination using built in functions in phyoseq (calls vegan)    </span></span>
<span id="cb57-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb57-2" aria-hidden="true" tabindex="-1"></a>ord.nmds.bray <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(ps, <span class="at">method=</span><span class="st">&quot;NMDS&quot;</span>,<span class="at">k=</span><span class="dv">2</span>, <span class="at">distance=</span><span class="st">&quot;bray&quot;</span>)</span></code></pre></div>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.04869473 
## Run 1 stress 0.0478881 
## ... New best solution
## ... Procrustes: rmse 0.01639951  max resid 0.04610477 
## Run 2 stress 0.04346713 
## ... New best solution
## ... Procrustes: rmse 0.05294335  max resid 0.1316673 
## Run 3 stress 0.05216743 
## Run 4 stress 0.04346694 
## ... New best solution
## ... Procrustes: rmse 0.0006780795  max resid 0.002148741 
## ... Similar to previous best
## Run 5 stress 0.05216581 
## Run 6 stress 0.04346719 
## ... Procrustes: rmse 0.0007087707  max resid 0.002245379 
## ... Similar to previous best
## Run 7 stress 0.04788334 
## Run 8 stress 0.05218892 
## Run 9 stress 0.04346725 
## ... Procrustes: rmse 0.0002716828  max resid 0.0008608943 
## ... Similar to previous best
## Run 10 stress 0.04346722 
## ... Procrustes: rmse 0.0007270832  max resid 0.0023034 
## ... Similar to previous best
## Run 11 stress 0.04788122 
## Run 12 stress 0.04346737 
## ... Procrustes: rmse 0.0008170541  max resid 0.002587991 
## ... Similar to previous best
## Run 13 stress 0.04788857 
## Run 14 stress 0.043467 
## ... Procrustes: rmse 6.761702e-05  max resid 0.0002142312 
## ... Similar to previous best
## Run 15 stress 0.05217513 
## Run 16 stress 0.05218083 
## Run 17 stress 0.04346707 
## ... Procrustes: rmse 0.000608542  max resid 0.001927707 
## ... Similar to previous best
## Run 18 stress 0.1096243 
## Run 19 stress 0.05584216 
## Run 20 stress 0.04950525 
## *** Solution reached</code></pre>
<p>You should see each iteration of the NMDS until a solution is reached (i.e., stress was minimized after some number of reconfigurations of the points in 2 dimensions). You can increase the number of default iterations using the argument trymax=. which may help alleviate issues of non-convergence. If high stress is your problem, increasing the number of dimensions to k=3 might also help.</p>
<p>Often in ecological research, we are interested not only in comparing univariate descriptors of communities, like diversity (above), but also in how the constituent species — or the composition — changes from one community to the next.</p>
<p>One common tool to do this is non-metric multidimensional scaling, or NMDS. The goal of NMDS is to collapse information from multiple dimensions (e.g, from multiple communities, sites, etc.) into just a few, so that they can be visualized and interpreted.</p>
<p>Consider a single axis representing the abundance of a single species. Along this axis, we can plot the communities in which this species appears, based on its abundance within each.</p>
<p>Now consider a second axis of abundance, representing another species. We can now plot each community along the two axes (Species 1 and Species 2).</p>
<p>Keep going and imagine how many axes we would have to produce!!!!</p>
<p>The goal of NMDS is to represent the original position of communities in multidimensional space as accurately as possible using a reduced number of dimensions that can be easily plotted and visualized (and to spare your thinker).</p>
<p>NMDS does not use the absolute abundances of species in communities, but rather their rank orders. The use of ranks omits some of the issues associated with using absolute distance (e.g., sensitivity to transformation), and as a result is a flexible technique</p>
<p>We can use the function plot.ordination to use our NMDS scaling to position our samples on a ggplot</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb59-1" aria-hidden="true" tabindex="-1"></a>ord1<span class="ot">&lt;-</span><span class="fu">plot_ordination</span>(ps, ord.nmds.bray, <span class="at">color=</span><span class="st">&quot;char&quot;</span>, <span class="at">title=</span><span class="st">&quot;Bray NMDS&quot;</span>)</span>
<span id="cb59-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb59-2" aria-hidden="true" tabindex="-1"></a>ord1</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-75-1.png" width="672" /></p>
<p>And we can add customisations to improve the plot</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-1" aria-hidden="true" tabindex="-1"></a>ord1 <span class="sc">+</span> <span class="fu">stat_ellipse</span>(<span class="at">type=</span><span class="st">&quot;norm&quot;</span>,</span>
<span id="cb60-2"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">geom=</span><span class="st">&quot;polygon&quot;</span>,</span>
<span id="cb60-3"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-3" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">aes</span>(<span class="at">fill=</span>char),</span>
<span id="cb60-4"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">alpha=</span><span class="fl">0.3</span>,</span>
<span id="cb60-5"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb60-6"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-6" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">pch=</span><span class="dv">21</span>, <span class="fu">aes</span>(<span class="at">fill=</span>char), <span class="at">colour=</span><span class="st">&quot;white&quot;</span>, <span class="at">stroke=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">3</span>)<span class="sc">+</span></span>
<span id="cb60-7"><a href="introduction-to-statistical-analysis-of-the-microbiome-data.html#cb60-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-76-1.png" width="672" /></p>
</div>
</div>
<div id="question" class="section level2" number="4.6">
<h2><span class="header-section-number">4.6</span> Question</h2>
<p>We have now produced several different visualisations and analyses of the microbiome communities from these deep sea samples.
What do you think are the most important or interesting findings here?</p>
</div>
<div id="summary-3" class="section level2" number="4.7">
<h2><span class="header-section-number">4.7</span> Summary</h2>
<p>This is a <em>very</em> brief introduction to using <strong>big</strong> data. Hopefully you can see that many of the command line tools, data cleaning and analysis and visualisation you have been practising prepare you for Big Data Biology.</p>
<p>In this example, in order to have generated and processed this data, you would have needed to know</p>
<ul>
<li>Sampling techniques</li>
<li>Molecular lab skills</li>
<li>Data analysis and coding knowledge</li>
<li>Ecological skills (community analysis)</li>
</ul>
<p>Biology is increasingly interdisciplinary - and the skills you learn in one module should be transferable to other things you do.</p>
<p>And remember at the heart of everything is Data! It doesn’t matter what you are studying or why - good data skills are fundamental.</p>
<p>Thank you!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="ngs-sequence-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="applications.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-method.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
